processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java

```diff
@@ -51,25 +62,51 @@ public SelectorDimFilter(
   {
     byte[] dimensionBytes = StringUtils.toUtf8(dimension);
     byte[] valueBytes = (value == null) ? new byte[]{} : StringUtils.toUtf8(value);
+    byte[] extractionFnBytes = extractionFn == null ? new byte[0] : extractionFn.getCacheKey();
 
-    return ByteBuffer.allocate(2 + dimensionBytes.length + valueBytes.length)
+    return ByteBuffer.allocate(3 + dimensionBytes.length + valueBytes.length + extractionFnBytes.length)
                      .put(DimFilterCacheHelper.SELECTOR_CACHE_ID)
                      .put(dimensionBytes)
                      .put(DimFilterCacheHelper.STRING_SEPARATOR)
                      .put(valueBytes)
+                     .put(DimFilterCacheHelper.STRING_SEPARATOR)
+                     .put(extractionFnBytes)
                      .array();
   }
 
   @Override
   public DimFilter optimize()
   {
+    if (this.getExtractionFn() instanceof LookupExtractionFn
+        && ((LookupExtractionFn) this.getExtractionFn()).isOptimize()) {
+      LookupExtractionFn exFn = (LookupExtractionFn) this.getExtractionFn();
+
+      // If retainMissingValues is true, we can't do an unapply()-based optimization, there may be
+      // row values that match the selector value but are not included in the lookup map.
+      if (exFn.isRetainMissingValue()) {
+        return this;
```

Comment:
@gianm updated optimization for SelectorDimFilter, it will now add the selector value to the InDimFilter when optimizing with retainMissingValues (or return a single SelectorFilter if the reverse lookup was empty)